File: c/ray3.c
#include <stdio.h>
#include <stdlib.h>
#include "ray3.h"

ray3* ray3_new(vec3* origin, vec3* direction) {
    ray3* r = malloc(sizeof(ray3));
    if (r == NULL) {
        fprintf(stderr, "Out of memory\n");
        exit(1);
    }
    r->origin = origin;
    r->direction = direction;
    return r;
}

vec3* ray3_point_at_parameter(vec3* result, ray3* r, float t) {
    vec3* scaled_direction = vec3_scale(vec3_new(0, 0, 0), r->direction, t);
    result = vec3_add(result, r->origin, scaled_direction);
    free(scaled_direction);
    return result;
}

File: python/ray3.py
from vec3 import Vec3

class Ray3():
    def __init__(self, origin: Vec3, direction: Vec3):
        self.origin = origin
        self.direction = direction
    
    def point_at_parameter(self, t: float) -> Vec3:
        return self.origin + (self.direction * t)